File: rails_app/app/models/code_file.rb
# == Schema Information
#
# Table name: code_files
#
#  website_id            :integer
#  path                  :string
#  content               :string
#  content_tsv           :tsvector
#  shasum                :string
#  file_specification_id :integer
#  source_type           :text
#  source_id             :integer
#  created_at            :datetime
#  updated_at            :datetime
#

class CodeFile < ApplicationRecord
# This is a read-only model backed by a database view
# The view merges template_files and website_files
#
  self.table_name = 'code_files'
  self.primary_key = nil # View doesn't have a primary key
  
  # Default ordering since the view has no primary key
  default_scope { order(created_at: :desc, path: :asc) }
  
  def readonly?
    true
  end
  
  belongs_to :website
  
  belongs_to :source_file, polymorphic: true, optional: true, 
    foreign_type: :source_type, foreign_key: :source_id
  
  # Scopes for full-text search
  scope :search, ->(query) {
    where("content_tsv @@ plainto_tsquery('english', ?)", query)
  }
  
  scope :search_with_rank, ->(query) {
    sanitized = connection.quote(query)
    search(query)
      .reselect("code_files.*, ts_rank(content_tsv, plainto_tsquery('english', #{sanitized})) AS rank")
      .order("rank DESC")
  }
  
  scope :search_phrase, ->(phrase) {
    where("content_tsv @@ phraseto_tsquery('english', ?)", phrase)
  }
  
  scope :search_boolean, ->(query) {
    where("content_tsv @@ to_tsquery('english', ?)", query)
  }
  
  # Fuzzy path search using trigrams
  def self.path_similar_to(path, threshold = 0.3)
    sanitized_path = connection.quote(path)
    
    # Use a subquery to avoid conflicts with default scope
    from(
      unscoped.select("code_files.*", "similarity(path, #{sanitized_path}) AS path_similarity"),
      :code_files
    ).where("similarity(path, ?) > ?", path, threshold)
     .order("path_similarity DESC")
  end
  
  scope :path_fuzzy, ->(pattern) {
    where("path % ?", pattern)
  }
  
  # Filter by source (website or template)
  scope :from_website, -> { where(source_type: 'WebsiteFile') }
  scope :from_template, -> { where(source_type: 'TemplateFile') }
  
  # Filter by polymorphic source
  scope :for_source, ->(source_type, source_id) { where(source_type: source_type, source_id: source_id) }
  scope :for_website_file, ->(website_file_id) { where(source_type: 'WebsiteFile', source_id: website_file_id) }
  scope :for_template_file, ->(template_file_id) { where(source_type: 'TemplateFile', source_id: template_file_id) }
  
  # Filter by file type
  scope :javascript, -> { where("path ILIKE '%.js' OR path ILIKE '%.jsx'") }
  scope :typescript, -> { where("path ILIKE '%.ts' OR path ILIKE '%.tsx'") }
  scope :css, -> { where("path ILIKE '%.css' OR path ILIKE '%.scss' OR path ILIKE '%.sass'") }
  scope :html, -> { where("path ILIKE '%.html' OR path ILIKE '%.htm'") }
  
  # Search with highlighting
  def self.search_with_highlights(query, options = {})
    start_sel = options[:start_sel] || '<mark>'
    stop_sel = options[:stop_sel] || '</mark>'
    max_words = options[:max_words] || 20
    
    sanitized_query = connection.quote(query)
    highlight_col = "ts_headline('english', content, plainto_tsquery('english', #{sanitized_query}), 'StartSel=#{start_sel}, StopSel=#{stop_sel}, MaxWords=#{max_words}, MinWords=10') AS highlighted_content"
    
    # Use reselect to override any default select
    search(query)
      .reselect("code_files.*", highlight_col)
  end
  
  # Get top matching files with context  
  def self.search_with_context(query, limit = 10)
    sanitized_query = connection.quote(query)
    highlight_col = "ts_headline('english', content, plainto_tsquery('english', #{sanitized_query}), 'MaxWords=20, MinWords=10') AS highlighted_content"
    
    # Use reselect to override any default select
    search(query)
      .reselect("code_files.*", highlight_col)
      .limit(limit)
  end
  
  # Count matches per website
  def self.count_matches_by_website(query)
    search(query)
      .unscope(:order)  # Remove default ordering for grouped queries
      .group(:website_id)
      .count
  end
  
  def website_file?
    source_type == 'WebsiteFile'
  end
  
  def template_file?
    source_type == 'TemplateFile'
  end
  
  def source
    case source_type
    when 'WebsiteFile'
      'website'
    when 'TemplateFile'
      'template'
    else
      source_type
    end
  end
  
  def source_file_object
    case source_type
    when 'WebsiteFile'
      WebsiteFile.find_by(id: source_id)
    when 'TemplateFile'
      TemplateFile.find_by(id: source_id)
    end
  end
  
  def file_type
    File.extname(path).delete('.').downcase
  end
  
  def language
    case file_type
    when 'js', 'jsx' then 'javascript'
    when 'ts', 'tsx' then 'typescript'  
    when 'rb' then 'ruby'
    when 'py' then 'python'
    when 'css', 'scss', 'sass' then 'css'
    when 'html', 'htm' then 'html'
    else file_type
    end
  end
end

File: langgraph_app/app/models/codeFile.ts
import { BaseView, sql, and, eq, or } from "./base";
import { codeFileSchema, CodeFileSourceEnum } from "@types";
import { codeFiles as codeFilesView } from "app/db";
import { db } from 'app/db';
import { SQL } from 'drizzle-orm';

// A "WebsiteFile" is distinguished from a "TemplateFile":
//
// TemplateFile: Boilerplate — A file that is part of the template. Aka a default package.json, or tsconfig.json
// WebsiteFile: User-Modified — A modified template file, or new page. Aka a user-modified package.json, or components/Hero.tsx, or pages/IndexPage.tsx
//
// "CodeFile" is a view that represents ALL files for a given website, whether they've been modified or not. 
// You should use WebsiteFile when you need to update user-modified files, and CodeFile for reading/querying files with full text search.
export class CodeFileModel extends BaseView<typeof codeFilesView, typeof codeFileSchema> {
  protected static entity = codeFilesView;
  protected static schema = codeFileSchema;

  /**
   * Get base select fields used by all queries that need custom selects
   * Ensures consistency across methods that need to add extra fields
   */
  private static getBaseSelectFields() {
    return {
      websiteId: this.entity.websiteId,
      path: this.entity.path,
      content: this.entity.content,
      contentTsv: this.entity.contentTsv,
      shasum: this.entity.shasum,
      fileSpecificationId: this.entity.fileSpecificationId,
      sourceType: this.entity.sourceType,
      sourceId: this.entity.sourceId,
      createdAt: this.entity.createdAt,
      updatedAt: this.entity.updatedAt,
    };
  }

  /**
   * Transform results to add computed 'source' field based on sourceType
   * Maps: 'WebsiteFile' → 'website', 'TemplateFile' → 'template'
   */
  private static transformResults(results: any[]): any[] {
    const resultsWithSource = results.map((r: any) => ({
      ...r,
      source: r.sourceType === 'WebsiteFile' ? 'website' : 'template'
    }));
    return resultsWithSource;
  }

  // Full-text search methods
  static async search(query: string, websiteId?: number) {
    let baseQuery = db
      .select()
      .from(this.entity)
      .where(
        sql`${this.entity.contentTsv} @@ plainto_tsquery('english', ${query})`
      );

    if (websiteId) {
      baseQuery = baseQuery.where(
        and(
          sql`${this.entity.contentTsv} @@ plainto_tsquery('english', ${query})`,
          eq(this.entity.websiteId, websiteId)
        )
      ) as any;
    }

    const results = await baseQuery;
    return this.transformResults(results);
  }

  // Search with ranking
  static async searchWithRank(query: string, websiteId?: number, limit = 10, pathFilter?: string) {
    const tsQuery = sql`plainto_tsquery('english', ${query})`;
    const rankExpr = sql`ts_rank(${this.entity.contentTsv}, ${tsQuery})`;
    
    let conditions = sql`${this.entity.contentTsv} @@ ${tsQuery}`;
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }
    if (pathFilter) {
      conditions = and(conditions, sql`${this.entity.path} ILIKE ${'%' + pathFilter + '%'}`)!;
    }
    
    const results = await db
      .select({
        ...this.getBaseSelectFields(),
        rank: sql<number>`${rankExpr}`
      })
      .from(this.entity)
      .where(conditions)
      // Order by rank first, then by path for consistent ordering when ranks are equal
      .orderBy(sql`${rankExpr} DESC`, this.entity.path)
      .limit(limit);

    return this.transformResults(results);
  }

  // Phrase search
  static async searchPhrase(phrase: string, websiteId?: number, limit = 10, pathFilter?: string) {
    const tsQuery = sql`phraseto_tsquery('english', ${phrase})`;
    // Use ts_rank_cd for better phrase ranking
    const rankExpr = sql`ts_rank_cd(${this.entity.contentTsv}, ${tsQuery})`;
    
    let conditions = sql`${this.entity.contentTsv} @@ ${tsQuery}`;
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }
    if (pathFilter) {
      conditions = and(conditions, sql`${this.entity.path} ILIKE ${'%' + pathFilter + '%'}`)!;
    }
    
    const baseQuery = db
      .select({
        ...this.getBaseSelectFields(),
        rank: sql<number>`${rankExpr}`
      })
      .from(this.entity)
      .where(conditions)
      // Order by rank first, then by path for consistent ordering when ranks are equal
      .orderBy(sql`${rankExpr} DESC`, this.entity.path)
      .limit(limit);

    const results = await baseQuery;
    return this.transformResults(results);
  }

  // Boolean search (supports AND, OR, NOT operators)
  static async searchBoolean(query: string, websiteId?: number, limit = 10, pathFilter?: string) {
    const tsQuery = sql`to_tsquery('english', ${query})`;
    // Use ts_rank_cd (cover density) for better boolean query ranking
    const rankExpr = sql`ts_rank_cd(${this.entity.contentTsv}, ${tsQuery})`;
    
    let conditions = sql`${this.entity.contentTsv} @@ ${tsQuery}`;
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }
    if (pathFilter) {
      conditions = and(conditions, sql`${this.entity.path} ILIKE ${'%' + pathFilter + '%'}`)!;
    }
    
    const baseQuery = db
      .select({
        ...this.getBaseSelectFields(),
        rank: sql<number>`${rankExpr}`
      })
      .from(this.entity)
      .where(conditions)
      // Order by rank first, then by path for consistent ordering when ranks are equal
      .orderBy(sql`${rankExpr} DESC`, this.entity.path)

    const results = await baseQuery.limit(limit);
    return this.transformResults(results);
  }

  // Path-based search using trigram similarity
  static async pathSimilarTo(path: string, websiteId?: number, threshold = 0.3) {
    const similarityExpr = sql`similarity(${this.entity.path}, ${path})`;
    
    let conditions = sql`${similarityExpr} > ${threshold}`;
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }

    const results = await db
      .select({
        ...this.getBaseSelectFields(),
        pathSimilarity: sql<number>`${similarityExpr}`
      })
      .from(this.entity)
      .where(conditions)
      .orderBy(sql`${similarityExpr} DESC`);

    return this.transformResults(results);
  }

  // Fuzzy path search using ILIKE for pattern matching
  static async pathFuzzy(pattern?: string, websiteId?: number, limit?: number) {
    let conditions;
    if (!pattern) {
      conditions = sql`1=1`;
    } else {
      conditions = sql`${this.entity.path} ILIKE ${'%' + pattern + '%'}`;
    }
    
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }
    
    let baseQuery = db
      .select()
      .from(this.entity)
      .where(conditions)
      .orderBy(this.entity.path);
    
    if (limit) {
      baseQuery = baseQuery.limit(limit);
    }

    const results = await baseQuery;
    return this.transformResults(results);
  }

  // Search with highlighted results
  static async searchWithHighlights(
    query: string, 
    websiteId?: number,
    options: {
      startSel?: string;
      stopSel?: string;
      maxWords?: number;
      minWords?: number;
    } = {}
  ) {
    const { 
      startSel = '<mark>', 
      stopSel = '</mark>', 
      maxWords = 20,
      minWords = 10 
    } = options;

    const tsQuery = sql`plainto_tsquery('english', ${query})`;
    
    let conditions = sql`${this.entity.contentTsv} @@ ${tsQuery}`;
    if (websiteId) {
      conditions = and(conditions, eq(this.entity.websiteId, websiteId))!;
    }

    const results = await db
      .select({
        ...this.getBaseSelectFields(),
        highlightedContent: sql<string>`
          ts_headline(
            'english', 
            ${this.entity.content}, 
            ${tsQuery},
            'StartSel=${startSel}, StopSel=${stopSel}, MaxWords=${maxWords}, MinWords=${minWords}'
          )`
      })
      .from(this.entity)
      .where(conditions);

    return this.transformResults(results);
  }

  // Get files by type
  static async byFileType(websiteId: number, extensions: string[]) {
    const conditions = extensions.map(ext => 
      sql`${this.entity.path} ILIKE ${'%.' + ext}`
    );

    const results = await db
      .select()
      .from(this.entity)
      .where(
        and(
          eq(this.entity.websiteId, websiteId),
          or(...conditions)
        )
      );

    return results;
  }

  // Scope methods
  static async fromWebsite(websiteId: number) {
    return this.where({ 
      websiteId, 
      source: CodeFileSourceEnum.Website 
    });
  }

  static async fromTemplate(websiteId: number) {
    return this.where({ 
      websiteId, 
      source: CodeFileSourceEnum.Template 
    });
  }

  // Get JavaScript/TypeScript files
  static async javascriptFiles(websiteId: number) {
    return this.byFileType(websiteId, ['js', 'jsx']);
  }

  static async typescriptFiles(websiteId: number) {
    return this.byFileType(websiteId, ['ts', 'tsx']);
  }

  static async cssFiles(websiteId: number) {
    return this.byFileType(websiteId, ['css', 'scss', 'sass']);
  }

  static async htmlFiles(websiteId: number) {
    return this.byFileType(websiteId, ['html', 'htm']);
  }

  // Count matches by website
  static async countMatchesByWebsite(query: string) {
    const results = await db
      .select({
        websiteId: this.entity.websiteId,
        count: sql<number>`count(*)`
      })
      .from(this.entity)
      .where(
        sql`${this.entity.contentTsv} @@ plainto_tsquery('english', ${query})`
      )
      .groupBy(this.entity.websiteId);

    return results;
  }

  // Instance methods
  get isWebsiteFile(): boolean {
    return this.data.source === CodeFileSourceEnum.Website;
  }

  get isTemplateFile(): boolean {
    return this.data.source === CodeFileSourceEnum.Template;
  }

  get fileType(): string {
    const match = this.data.path.match(/\.([^.]+)$/);
    return match ? match[1].toLowerCase() : '';
  }

  get language(): string {
    const ext = this.fileType;
    const languageMap: Record<string, string> = {
      'js': 'javascript',
      'jsx': 'javascript',
      'ts': 'typescript',
      'tsx': 'typescript',
      'rb': 'ruby',
      'py': 'python',
      'css': 'css',
      'scss': 'css',
      'sass': 'css',
      'html': 'html',
      'htm': 'html'
    };
    return languageMap[ext] || ext;
  }
}